plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/preferences/RunOptionsDefaultsComponent.java

```diff
@@ -233,93 +330,105 @@ public void setEnabled(boolean enabled) {
+      if (Objects.equals(project, fetchStagingLocationsJob.getProject())
+          && Objects.equals(selectedEmail, fetchStagingLocationsJob.getAccountEmail())
+          && fetchStagingLocationsJob.getState() == Job.RUNNING) {
+        // an update is in progress
+        return;
+      }
+      fetchStagingLocationsJob.cancel();
+    }
+    fetchStagingLocationsJob = null;
     if (!Strings.isNullOrEmpty(project) && credential != null) {
-      ListenableFutureProxy<SortedSet<String>> stagingLocationsFuture =
-          FetchStagingLocationsJob.schedule(getGcsClient(), project);
-      UpdateStagingLocationComboListener updateComboListener =
-          new UpdateStagingLocationComboListener(stagingLocationsFuture);
-      stagingLocationsFuture.addListener(updateComboListener, executor);
+      final FetchStagingLocationsJob thisJob = fetchStagingLocationsJob =
+          new FetchStagingLocationsJob(getGcsClient(), selectedEmail, project);
+      fetchStagingLocationsJob.getStagingLocations().addListener(new Runnable() {
+        @Override
+        public void run() {
+          // check that this is same job (may have been cancelled in the interim)
+          if (!target.isDisposed() && fetchStagingLocationsJob == thisJob) {
+            // Update the Combo with the staging locations retrieved by the Job.
+            try {
+              SortedSet<String> stagingLocations = fetchStagingLocationsJob.getStagingLocations().get();
+              // Don't use "removeAll()", as it will clear the text field too.
+              stagingLocationInput.remove(0, stagingLocationInput.getItemCount() - 1);
+              for (String location : stagingLocations) {
+                stagingLocationInput.add(location);
+              }
+              completionListener.setContents(stagingLocations);
+            } catch (InterruptedException | ExecutionException ex) {
+              // ignore: handled by validate()
+            }
+            validate();
+          }
+        }
+      }, displayExecutor);
+      fetchStagingLocationsJob.schedule(scheduleDelay);
     }
   }
 
   /**
-   * Ensure the staging location specified in the input combo is valid.
+   * Return the staging location verification result future, or null if no job is yet scheduled.
    */
-  private void verifyStagingLocation(final String stagingLocation) {
-    setPageComplete(false);
-    messageTarget.clear();
-
-    if (verifyJob != null) {
-      // Cancel any existing verifyJob
-      // FIXME: this has no effect, as "VerifyStagingLocationJob" doesn't honor cancellation.
-      verifyJob.cancel();
-    }
-
-    final String bucketNamePart = extractBucketNamePart();
-    if (bucketNamePart.isEmpty()) {
-      // If the bucket name is empty, we don't have anything to verify; and we don't have any
-      // interesting messaging.
-      setPageComplete(true);
-      return;
-    }
+  @VisibleForTesting
+  synchronized Future<VerifyStagingLocationResult> getVerifyStagingLocationResult() {
```

Comment:
This method is called only in the UI thread, so `synchronized` is not necessary. Please see my previous comment.